### ISSUE TITLE
shinydashboard error: "Unable to find input binding for element..."

### ISSUE THREAD

**Issue by lbramley** [rstudio/shinytest#171]
I'm getting an error when testing a shinydashboard app that splits inputs across multiple tabs: `Error in session_makeRequest(self, private, endpoint, data, params, headers) : 
  Unable to find input binding for element with id selectFruit`

For example, the following works with `recordTest()` and `testApp()`:

```
library(shiny)
library(shinydashboard)

header <- dashboardHeader()

sidebar <- dashboardSidebar()

body <- dashboardBody(textInput("newFruit", label = "What fruit do you want to add?",
                                placeholder = "Enter name here"),
                      actionButton("saveBtn", "Save Fruit to List"),
                      uiOutput("selectUI"),
                      textOutput("result"))

ui <- dashboardPage(header, sidebar, body)

server <- function(input, output, session) {
  
  rv <- reactiveValues(fruit = c("", "Apple", "Banana", "Kiwi"))
  
  observeEvent(input$saveBtn, {
    
    if(input$saveBtn != 0) {
      rv$fruit <- c(rv$fruit, input$newFruit)
    }
  })
  
  output$selectUI <- renderUI({
    list(
      selectInput("selectFruit", label = "Which fruit is the best?", 
                  choices = rv$fruit)
    )
  })
  
  output$result <- renderText({
    req(input$selectFruit)
    paste0("You selected ", input$selectFruit, ". You're right, that's the best!")
  })
}

shinyApp(ui, server)
```

But if you instead split inputs across two tabs by replacing the `sidebar` and `body` objects with the following, it will throw the error.

```
sidebar <- dashboardSidebar(
  sidebarMenu(
    menuItem("Add fruit options", tabName = "addFruit"),
    menuItem("ChooseAFruit", tabName = "chooseFruit")
  )
)

body <- dashboardBody(
  tabItems(
    tabItem("addFruit",
            textInput("newFruit", label = "What fruit do you want to add?",
                      placeholder = "Enter name here"),
            actionButton("saveBtn", "Save Fruit to List")),
    tabItem("chooseFruit",
            uiOutput("selectUI"),
            textOutput("result"))
  )
)
```

**Comment by JBAkroyd** [comment]
Getting the same errors

```
> testApp("/srv/shiny-server/joakr0/Genomic_Reporting/genrep/R", "tableViewCohortYear")
Running tableViewCohortYear.R Error in session_makeRequest(self, private, endpoint, data, params, headers) : 
  Unable to find input binding for element with id sidebarCollapsed
> testApp("/srv/shiny-server/joakr0/Genomic_Reporting/genrep/R", "tableViewCohortYear")
Running tableViewCohortYear.R Error in session_makeRequest(self, private, endpoint, data, params, headers) : 
  Unable to find input binding for element with id sidebarItemExpanded
> testApp("/srv/shiny-server/joakr0/Genomic_Reporting/genrep/R", "tableViewCohortYear")
Running tableViewCohortYear.R Error in session_makeRequest(self, private, endpoint, data, params, headers) : 
  Unable to find input binding for element with id caRunDate1Input
```

Not only does the shinydashboard package itself not work with shinytest, it also mean it can't test variables without it.

**Comment by wch** [comment]
I believe this is a duplicate of #167, and it was fixed by rstudio/shinydashboard#256.

**Comment by felipegerard** [comment]
Hi! I'm having a similar problem when using shinydashboard + submodules. For some reason shinytest is unable to find elements produced by submodules.

app.R
```
#
# This is a Shiny web application. You can run the application by clicking
# the 'Run App' button above.
#
# Find out more about building applications with Shiny here:
#
#    http://shiny.rstudio.com/
#

library(shiny)
library(shinydashboard)
library(shinytest)

## Submodule UI
subUI <- function(id) {
    ns <- shiny::NS(id)
    fluidPage(
        textInput(ns('text_input'), 'Write something!'),
        textOutput(ns('text_output'))
    )
    
}

## Submodule Server
subServer <- function(input, output, session) {
    output$text_output <- renderText({
        paste('You wrote this:', input$text_input)
    })
}

# UI
ui <- dashboardPage(
    dashboardHeader(title = 'Test'),
    dashboardSidebar(
        sidebarMenu(
            uiOutput('sidebar')
        )
    ),
    dashboardBody(
        uiOutput('body')
    )
)

# Server
server <- function(input, output, session) {
    
    logged_in <- reactiveVal(FALSE)
    observeEvent(input$login_button + input$logout_button, {
        logged_in(!logged_in())
    }, ignoreInit = TRUE)
    output$sidebar <- renderUI({
        if (logged_in()) {
            sidebarMenu(
                menuItem(tabName = 'test_item', 'Test Item')
            )
        } else {
            sidebarMenu(
                menuItem(tabName = 'login', 'Log in')
            )
        }
    })
    output$body <- renderUI({
        if (logged_in()) {
            tabItems(
                tabItem(
                    tabName = 'test_item',
                    uiOutput("sub_ui"),
                    actionButton('logout_button', 'Logout')
                )
            )
        } else {
            tabItems(
                tabItem(
                    tabName = 'login',
                    actionButton('login_button', 'Login')
                )
            )
        }
    })
    output$sub <- callModule(
        subServer,
        'sub'
    )
    
    output$sub_ui <- renderUI({
        subUI('sub')
    })
}

# Run the application 
shinyApp(ui = ui, server = server)
```

mytest.R
```
app <- ShinyDriver$new("../")
app$snapshotInit("mytest")

app$setInputs(sidebarCollapsed = FALSE)
app$setInputs(logout_button = "click")
app$setInputs(login_button = "click")
app$setInputs(login_button = "click")
app$snapshot()
app$setInputs(`sub-text_input` = "asdf") #this doesn't get found
app$snapshot()
app$setInputs(logout_button = "click")
app$snapshot()
```

The error reads:
```
Running mytest.R Error in session_makeRequest(self, private, endpoint, data, params, headers) : 
  Unable to find input binding for element with id sub-text_input
```

Any ideas?
